Readme.md
# swift-everywhere-toolchain

## Requirements

- macOS 11.x
- Xcode 12.x
- Android Studio 4.1
- Android NDK 20.1.5948944
- Node 12.18 (node -v). Newer versions may also work, but not tested.
- CMake 3.19.X (cmake --version)
- Ninja 1.10 (ninja --version)

## Important

- Toolchain build may fail if macOS headers installed under `/usr/include`. This usually happens if you previously installed package `/Library/Developer/CommandLineTools/Packages/macOS_SDK_headers_for_macOS_10.14.pkg`. See details in [Xcode Command Line Tools](https://developer.apple.com/documentation/xcode_release_notes/xcode_10_release_notes#3035624) notes. See following [SuperUser question](https://superuser.com/questions/36567/how-do-i-uninstall-any-apple-pkg-package-file) about how to uninstall package.
- Toolchain build may fail if the command line tools are present in `/Library/Developer/CommandLineTools`. Remove them if you are not using them. See: [macos - How do I uninstall the command line tools for Xcode? - Ask Different](https://apple.stackexchange.com/questions/308943/how-do-i-uninstall-the-command-line-tools-for-xcode)

Keep tools like `CMake` and `ninja` up to date.

## Using pre-built toolchain

Build of complete toolchain takes ~1.5h. Instead of building it you can just download and use already pre-built package from [Releases](https://github.com/vgorloff/swift-everywhere-toolchain/releases) page.

## Setup and Build

1. Check that CMake is installed. Install CMake if needed.

   ```bash
   which cmake
   brew install cmake
   ```

2. Check that Ninja is installed. Install Ninja if needed.

   ```bash
   which ninja
   brew install ninja
   ```

3. Check that Autotools is installed. Install Autotools if needed.

   ```bash
   which autoconf
   brew install autoconf

   which aclocal
   brew install automake

   which glibtool
   brew install libtool

   which pkg-config
   brew install pkg-config
   ```

4. Check that git-lfs is installed. Install git-lfs if needed.

   ```bash
   which git-lfs
   brew install git-lfs
   ```

5. Make sure that `Xcode Build Tools` properly configured.

   ```bash
   xcode-select --print-path
   ```

6. Clone this repository.

   ```bash
   git clone https://github.com/vgorloff/swift-everywhere-toolchain.git
   cd swift-everywhere-toolchain
   ```

7. Create a symbolic link to NDK installation directory.

   ```bash
   sudo mkdir -p /usr/local/ndk
   sudo ln -vsi ~/Library/Android/sdk/ndk/$VERSION /usr/local/ndk/$VERSION
   ```

   The placeholder `$VERSION` needs to be replaced with a version mentioned in file `NDK_VERSION` at the root of cloned repository.

8. Start a build.

   ```bash
   node main.js
   ```

9. Once the build completed, toolchain will be saved to folder `ToolChain/swift-android-toolchain` and compressed into archive `ToolChain/swift-android-toolchain.tar.gz`.

## Usage

Please refer file [Assets/Readme.md](Assets/Readme.md) to see how to compile Swift files or build Swift packages using Toolchain.

## Sample Projects

Sample projects can be found in a separate [swift-everywhere-samples](https://github.com/vgorloff/swift-everywhere-samples) repository. Please look into `Readme.md` in that repository to get information about how to configure and build sample projects.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.65] - 2020-11-29
* [#102] Allow to provide custom -D__ANDROID_API__ definition.

## [1.0.64] - 2020-11-28
* [#99] Used NDK API level downgraded to 21 (__ANDROID_API_M__)

## [1.0.63] - 2020-11-28
* [#97] Used NDK API level downgraded to 23 (__ANDROID_API_M__)

## [1.0.62] - 2020-11-27
* [#95] Swift and related components updated to v.5.3.1

## [1.0.61] - 2020-11-26
* [#92] Android NDK updated to v.20.1.5948944

## [1.0.60] - 2020-11-23
* [#89] Fixes compile error when building C++ sources using SPM.

## [1.0.59] - 2020-11-22
* [#89] Fixed wrong LC_RPATH settings for spm libraries.

## [1.0.58] - 2020-11-17
### Changed
* Fixed installer to copy forgotten licenses for llbuild and spm.
* Fixed SPMBuilder to copy forgotten libraries.

## [1.0.57] - 2020-11-14
### Added
+ SPM and LLB Builders.
+ Toolchain now contains `swift-build` binary.
### Changed
* Fixed error "unknown argument: '-modulewrap'" when building sources using Swift Package Manager.

## [1.0.56] - 2020-11-10
### Changed
* Build scripts converted from Ruby to JS
* Introduced Swift 5.3 support.

## [1.0.55] - 2020-06-06
### Changed
* Updated libFoundation
### Fixed
* Fixes Swift 5.2 checkout.
* Fixes libFoundation link error due unknown option `-pthread` passed to swiftc.

## [1.0.54] - 2020-03-25
### Changed
* Swift updated to version 5.2.
* libDispatch, libFoundation, LLVM, Cmark updated to match Swift version 5.2.

## [1.0.53] - 2020-01-16
### Changed
* libFoundation switched to recent version.
* OpenSSL updated to OpenSSL_1_1_1d.
* CURL updated to curl-7_68_0.

## [1.0.52] - 2020-01-15
### Changed
* libDispatch switched to recent version.

## [1.0.51] - 2020-01-15
### Changed
* Switched to monolithic LLVM project.

## [1.0.50] - 2020-01-14
### Changed
* Updated LLVM dependencies to latest non-monolithic LLVM project.

## [1.0.49] - 2020-01-13
### Changed
* Updated to support NDK r20b.

## [1.0.48] - 2019-08-19
### Changed
* Updated compiler settings to fit recent changes in Swift.

## [1.0.47] - 2019-08-15
### Changed
* Swift updated to latest version.

## [1.0.46] - 2019-08-12
### Changed
* Scripting: Avoid copying files if destination up to date.
* libFoundation updated to latest version.

## [1.0.45] - 2019-08-09
### Added
+ Test targets to quickly test Toolchain after build.
### Fixed
* Fixed wrong paths in Module map file.

## [1.0.44] - 2019-08-08
### Added
+ Scripts which allow to use toolchain with Swift Package Manager from Xcode 11.
### Removed
- llbuild and Swift Package Manager now NOT a part or toolchain.

## [1.0.43] - 2019-08-06
### Added
+ llbuild and Swift Package Manager now part or toolchain.

## [1.0.42] - 2019-08-05
### Changed
* libDispatch and libFoundation updated to latest versions.

## [1.0.41] - 2019-08-04
### Fixed
* Fixed libFoundation build error when Cmake v.3.15.x is used.

## [1.0.40] - 2019-07-24
### Changed
* Updated Swift, libFoundation to latest revisions.

## [1.0.38] - 2019-07-16
### Changed
* Updated Dispatch to latest revision.

## [1.0.37] - 2019-07-10
### Changed
+ Added SPM and LLBuild builders. They are not used in toolchain yet.

## [1.0.36] - 2019-06-26
### Changed
- Removed patch as it now part of libFoundation. See: https://github.com/apple/swift-corelibs-foundation/pull/2376

## [1.0.35] - 2019-06-25
### Changed
* Replaced patch for `cmake/modules/AddSwift.cmake` according to PR review notes: https://github.com/apple/swift/pull/25682


Scripts/Common/Revision.rb
#
# The MIT License
#
# Copyright (c) 2019 Volodymyr Gorlov (https://github.com/vgorloff)
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.
#

class Revision

   # See also: $ToolChain/Sources/swift/utils/update_checkout/update-checkout-config.json

   def self.swift
      return "71d85a7c28eed8f46241649a723ddf23989139c6"
      # 1.0.55 @ 2020.05.18 @ origin/swift-5.2-branch: b623e823d42841ab8b151fb9c092bb0f9f5084e6
      # 1.0.54 @ 2020.03.24 @ swift-5.2-RELEASE: 443e9a41d1a7c1fcae280490144fcbf8461d3499 // This commit is gone now :0
      # 1.0.53 @ swift-DEVELOPMENT-SNAPSHOT-2020-01-13-a: 3b9014dc9acae3d95255a83befd85b52e7176acb
      # 1.0.51 @ swift-DEVELOPMENT-SNAPSHOT-2019-12-27-a: 746b58e8e1c1ef1ac09d8031875fd2a08b65597c
      # 1.0.50 @ swift-DEVELOPMENT-SNAPSHOT-2019-10-14-a: 9a4abf47a450aeff6a3231225fd2cf18cc7a940a
      # 1.0.50 @ swift-DEVELOPMENT-SNAPSHOT-2019-09-26-a: 8a91e9802448de399df4ee3ed24c63dc25af7e67
      # 1.0.50 @ swift-DEVELOPMENT-SNAPSHOT-2019-09-18-a: 4dac0d1e8bffee79728f04784961b3bc26ed9dbc
      # 1.0.49 @ swift-DEVELOPMENT-SNAPSHOT-2019-09-16-a: f302da0f9b856592992206afa75dd95668a192e1
      # 1.0.49 @ swift-DEVELOPMENT-SNAPSHOT-2019-08-30-a: dcd0893efe43639dddf6fc1637022822080e6a6a
   end

   def self.dispatch
      return "c992dacf3ca114806e6ac9ffc9113b19255be9fe"
      # 1.0.54 @ 2019.10.01 @ swift-5.2-RELEASE: c992dacf3ca114806e6ac9ffc9113b19255be9fe
      # 1.0.52 @ swift-DEVELOPMENT-SNAPSHOT-2020-01-13-a: c992dacf3ca114806e6ac9ffc9113b19255be9fe
      # 1.0.52 @ swift-DEVELOPMENT-SNAPSHOT-2019-09-29-a: 2accb0b97738b9ba9a0ffc1cb335716b70f9f65a
      # 1.0.49 @ swift-DEVELOPMENT-SNAPSHOT-2019-09-26-a: 970d562fcaee756bae9d20386e64bba90e635818
      # 1.0.49 @ swift-DEVELOPMENT-SNAPSHOT-2019-09-16-a: c023edd1cd02720407c62f4fac6a3f2c92cd0069
      # 1.0.49 @ swift-DEVELOPMENT-SNAPSHOT-2019-09-02-a: 90a45ce22f62c7c380339aa847dae4ab2d66ee30
      # 1.0.49 @ swift-DEVELOPMENT-SNAPSHOT-2019-08-22-a: 3da29ddcab09e4d09b015624e377f73a82c8dc9c
   end

   def self.foundation
      return "28267aafd965c3950a6d84fed1ec593257b82c58"
      # 1.0.54 @ 2020.03.18 @ swift-5.2-RELEASE: 8d0871530c52456278ef8c83cb049e6d70b28a6b
      # 1.0.53 @ swift-DEVELOPMENT-SNAPSHOT-2020-01-05-a: 62809e9e2f918e6c11e47cd6a8d858429c77b743
      # 1.0.53 @ swift-DEVELOPMENT-SNAPSHOT-2019-12-27-a: 86eec6148c1e4ef1c56dbd5e53c5cbf6d023f202
      # 1.0.53 @ swift-DEVELOPMENT-SNAPSHOT-2019-11-30-a: 2195121bf063e2db7c71080fcae7e76e8a43950e
      # 1.0.53 @ swift-DEVELOPMENT-SNAPSHOT-2019-11-06-a: c093eb65ed3aabb334857714e01989ecfb875120
      # 1.0.49 @ swift-DEVELOPMENT-SNAPSHOT-2019-11-01-a: facf5715d25d1765f10e4319b0701f300d9248e0
      # 1.0.49 @ swift-DEVELOPMENT-SNAPSHOT-2019-10-30-a: 5bb98287f5945a8cbb37df5a69741328488d2dd0
      # 1.0.49 @ swift-DEVELOPMENT-SNAPSHOT-2019-09-30-a: f62c026cb2f2f162bb3b3d93ae49969b60652e0b
      # 1.0.49 @ swift-DEVELOPMENT-SNAPSHOT-2019-09-04-a: c9c34a9447682131ae49332e3d2a20b8c82ca01e
      # 1.0.49 @ swift-DEVELOPMENT-SNAPSHOT-2019-09-02-a: 3e5366251756d7ffcf289c0faff865ca7dd23417
      # 1.0.49 @ swift-DEVELOPMENT-SNAPSHOT-2019-08-28-a: d99abed49a0763eefb7e1e1cdaa9654fb38c7c1b
      # 1.0.49 @ swift-DEVELOPMENT-SNAPSHOT-2019-08-16-a: 2e4734c674247b011a67996a9472f15ecb394e5e
      # 1.0.49 @ swift-DEVELOPMENT-SNAPSHOT-2019-08-14-a: f2d055cf77f59568c23412b2179e72193abac052
   end

   def self.spm
      # swift-5.2-branch @ 2020.02.28 @ 1.0.54 f2318da4f96b0607ade27eff51558206f57c2f98
      # swift-5.1-branch @ 1.0.50 331fa5cb245be1e4e16640f20f123eb3461eeced
      # master @ 1.0.43 3dad9e7cfe20d1b48256afca13df293bdca870b9
      # swift-5.0-branch @ 1.0.43 3a57975e10be0b1a8b87992ddf3a49701036f96c
      # swift-5.1-branch @ 1.0.37 0996526dbf2d5adbdef4b133312e441c0deb3db7
      return "59866b68f14c71a36c00eb6b2076eb120c21b0a6"
   end

   def self.llb
      # swift-5.2-branch @ 2020.01.07 @ 1.0.53 fb7ebf0b06c0d7c45ca8e18b3371424503a38b5c
      # swift-5.1-branch @ 1.0.50 f73b84bc1525998e5e267f9d830c1411487ac65e
      # master @ 1.0.43 5bf016632c0e85d835c63b84b43d3c887f01418f
      # swift-5.0-branch @ 1.0.43 4f77fc47e2f5fcddc057abc8391f7c23d1e8d275
      # swift-5.1-branch @ 1.0.37 f73b84bc1525998e5e267f9d830c1411487ac65e
      return "d6dd04202d509f91d61aef89236db0da8f27d356"
   end

   def self.llvm
      return "c4513204aeb003122d2c69bd454ab8781298453b"
      # swift-5.2.5-RELEASE: "c4513204aeb003122d2c69bd454ab8781298453b"
      # 1.0.54 @ 2020.03.23 @ swift-5.2-RELEASE: c2133f2b08047708b17230b598cb789062143650
      # 1.0.54 @ swift-5.1.5-RELEASE: 198fa1e3efcce2fd302f7aaa077aec443a82394b
      # 1.0.51 @ swift-DEVELOPMENT-SNAPSHOT-2020-01-03-a: 019936b2100dc33c8946928d424b5cff4211acea
   end

   def self.cmark
      return "68e209466644ce07dcb0b695ea4d55aac4f0b564"
      # 1.0.54 @ 2019.10.10 @ swift-5.2-RELEASE: bfa95d55b535fa178f75484b5e8f82ae3d8517af
      # 1.0.51 @ swift-DEVELOPMENT-SNAPSHOT-2020-01-03-a: 31e12547c4de3dfcacc69855d93298cb90fd4f27
      # 1.0.50 @ swift-DEVELOPMENT-SNAPSHOT-2019-10-14-a: bfa95d55b535fa178f75484b5e8f82ae3d8517af
      # 1.0.49 @ swift-5.1.1-RELEASE: 32fa49671d0fc5d1f65d2bcbabfb1511a9d65c27
   end

   def self.ssl
      return "894da2fb7ed5d314ee5c2fc9fd2d9b8b74111596"
      # 1.0.53 @ OpenSSL_1_1_1d: 894da2fb7ed5d314ee5c2fc9fd2d9b8b74111596
      # OpenSSL_1_1_1c: 97ace46e11dba4c4c2b7cb67140b6ec152cfaaf4 (1.0.26)
      # OpenSSL_1_1_1b: 50eaac9f3337667259de725451f201e784599687
      # OpenSSL_1_1_1a: d1c28d791a7391a8dc101713cd8646df96491d03
   end

   def self.icu
      return "e2d85306162d3a0691b070b4f0a73e4012433444"
      # 1.0.26 @ v64-2: e2d85306162d3a0691b070b4f0a73e4012433444
      # v64-rc2: 67d218f2476ac543de2ed843fa080892972c604a
   end

   def self.curl
      return "2cfac302fbeec68f1727cba3d1705e16f02220ad"
      # 1.0.53 @ curl-7_68_0: 2cfac302fbeec68f1727cba3d1705e16f02220ad
      # 1.0.26 @ 7_65_1: 69248b58f649e35b09a126c12781353e3471f5c6
   end

   def self.xml
      # v2.9.9: f8a8c1f59db355b46962577e7b74f1a1e8149dc6 (1.0.26)
      return "f8a8c1f59db355b46962577e7b74f1a1e8149dc6"
   end

end
